lib/web_ui/dev/screenshot_manager.dart

```diff
@@ -0,0 +1,224 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.6
+import 'dart:io' as io;
+import 'dart:convert';
+import 'dart:math';
+
+import 'package:image/image.dart';
+import 'package:path/path.dart' as path;
+import 'package:webkit_inspection_protocol/webkit_inspection_protocol.dart'
+    as wip;
+import 'package:yaml/yaml.dart';
+
+import 'common.dart';
+import 'environment.dart';
+import 'safari_installation.dart';
+import 'utils.dart';
+
+/// [ScreenshotManager] implementation for Chrome.
+///
+/// This manager can be used for both macOS and Linux.
+// TODO: https://github.com/flutter/flutter/issues/65673
+class ChromeScreenshotManager extends ScreenshotManager {
+  String get filenameSuffix => '';
+
+  /// Capture a screenshot of the web content.
+  ///
+  /// Uses Webkit Inspection Protocol server's `captureScreenshot` API.
+  ///
+  /// [region] is used to decide which part of the web content will be used in
+  /// test image. It includes starting coordinate x,y as well as height and
+  /// width of the area to capture.
+  Future<Image> capture(Rectangle region) async {
+    final wip.ChromeConnection chromeConnection =
+        wip.ChromeConnection('localhost', kDevtoolsPort);
+    final wip.ChromeTab chromeTab = await chromeConnection.getTab(
+        (wip.ChromeTab chromeTab) => chromeTab.url.contains('localhost'));
+    final wip.WipConnection wipConnection = await chromeTab.connect();
+
+    Map<String, dynamic> captureScreenshotParameters = null;
+    if (region != null) {
+      captureScreenshotParameters = <String, dynamic>{
+        'format': 'png',
+        'clip': <String, dynamic>{
+          'x': region.left,
+          'y': region.top,
+          'width': region.width,
+          'height': region.height,
+          'scale':
+              // This is NOT the DPI of the page, instead it's the "zoom level".
+              1,
+        },
+      };
+    }
+
+    // Setting hardware-independent screen parameters:
+    // https://chromedevtools.github.io/devtools-protocol/tot/Emulation
+    await wipConnection
+        .sendCommand('Emulation.setDeviceMetricsOverride', <String, dynamic>{
+      'width': kMaxScreenshotWidth,
+      'height': kMaxScreenshotHeight,
+      'deviceScaleFactor': 1,
+      'mobile': false,
+    });
+    final wip.WipResponse response = await wipConnection.sendCommand(
+        'Page.captureScreenshot', captureScreenshotParameters);
+
+    final Image screenshot =
+        decodePng(base64.decode(response.result['data'] as String));
+
+    return screenshot;
+  }
+}
+
+/// [ScreenshotManager] implementation for Safari.
+///
+/// This manager is will only be created/used for macOS.
+class IOSSafariScreenshotManager extends ScreenshotManager {
```

Comment:
For consistency with other classes, the name of this class should begin with `Ios`, not `IOS`.

Reply:
done